cgp-utils_config.md

```diff
@@ -0,0 +1,114 @@
+
+To use the `cpg-utils.config` functions, import these two key methods into any script:
+
+```python
+from cpg_utils.config import get_config, set_config_paths
+```
+
+Configuration is loaded lazily, so config files are loaded into memory only when we try to retrieve values from them. This reduces start-up overhead for scripts using this config approach, but can result in late failures if files with invalid content are specified.
+
+### Setting configuration content
+
+Configuration files can be set in two ways:
+
+1. Set the `CPG_CONFIG_PATH` environment variable
+2. Use `set_config_paths` to point to one or more config TOMLs
+
+`CPG_CONFIG_PATH` can be a comma-delimited String, and the argument to `set_config_paths` is a list of `strings`. The
+config functionality allows for multiple files to be specified, applied in the following way:
+
+1. The cpg-utils repository contains a template config,
+[toml template](https://github.com/populationgenomics/cpg-utils/blob/main/cpg_utils/config-template.toml); this is loaded first and becomes the base config.
+2. For each additional config file in order, recursively update the base config with further content. New content is added, and content with the exact same key is updated/replaced.
+
+#### config loading example
+
+First file:
+
+```toml
+[file]
+name = "first.toml"
+[content]
+square = 4
+```
+
+Second file:
+
+```toml
+[file]
+name = "second.toml"
+[content]
+triangle = 3
+```
+
+Result:
+
+```python
+{
+    'file': 'second.toml',
+    'content': {
+        'square': 4,
+        'triangle': 3
+    }
+}
+```
+
+It's important to note that the config files are loaded 'left-to-right', so when multiple configuration files are loaded, only the right-most value for any overlapping keys will be retained.  
+
+### Getting configuration content
+
+Once configuration paths are set, values can be retrieved at any point using `get_config`. The first call to `get_config` sets the global config dictionary and returns the content, subsequent calls will just return the config dictionary.
+
+```python
+assert get_config()['file'] == 'second.toml'
+```
+
+## Analysis-Runner
+
+Analysis-runner incorporates a simple interface for config setting. When setting off a job, the flag `--config` can be used, pointing to a config file (local, or within GCP and accessible with current logged-in credentials).
+The `--config` flag can be used multiple times, which will cause the argument files to be aggregated in the order they are defined. When `--config` is set in this way, the job-runner performs the following actions:
+
+1. Locally (where `analysis-runner` is invoked), a [merged configuration file is generated](https://github.com/populationgenomics/analysis-runner/blob/main/analysis_runner/cli_analysisrunner.py#L199-L201), creating a single dictionary
+2. This dictionary is sent with the job definition to the execution server
+3. The data is saved in TOML format to a dataset-specific temporary location in GCP
```

Comment:
Might be worth mentioning that this is helpful for reproducible runs (as e.g. local configs that are not committed to `git` are also supported, as you already mention above). The path to the merged config is logged in Hail Batch, at least for `analysis-runner` invocations.

Reply:
We do store the merged configs permanently in `gs://cpg-config`: https://github.com/populationgenomics/analysis-runner/blob/ba826f574f1a1049ce01bc7236073bdbd3b3a0ec/server/util.py#L168-L173

Reply:
Ah sorry 🤦  I thought I remembered the file being written to -tmp! 🤦🤦